sdk/instrumentation/instrumentation-opentelemetry/test/internal/inMemoryInstrumenter.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  Instrumenter,
+  InstrumenterSpanOptions,
+  SpanStatus,
+  TracingContext,
+  TracingSpan,
+  TracingSpanContext
+} from "@azure/core-tracing";
+
+export class InMemoryTracingSpan implements TracingSpan {
+  name: string;
+  constructor(name: string) {
+    this.name = name;
+  }
+  status?: SpanStatus;
+  attributes: Record<string, unknown> = {};
+  endCalled: boolean = false;
+  exception?: string | Error;
+  setStatus(status: SpanStatus): void {
+    this.status = status;
+  }
+  setAttribute(name: string, value: unknown): void {
+    this.attributes[name] = value;
+  }
+  end(): void {
+    this.endCalled = true;
+  }
+  recordException(exception: string | Error): void {
+    this.exception = exception;
+  }
+  isRecording(): boolean {
+    return true;
+  }
+  get spanContext() {
+    return {
+      spanId: "",
+      traceFlags: 0,
+      traceId: ""
+    };
+  }
+}
+
+export class InMemoryInstrumenter implements Instrumenter {
+  public contexts: TracingContext[] = [new ContextImpl()];
+  public startedSpans: InMemoryTracingSpan[] = [];
+
+  startSpan(
+    name: string,
+    spanOptions?: InstrumenterSpanOptions
+  ): { span: TracingSpan; tracingContext: TracingContext } {
+    const span = new InMemoryTracingSpan(name);
```

Comment:
so we don't take the package name here? Is it like we set the initial span or something?

Reply:
Good point... we should figure out how we want folks to test that they passed the correct package name...

In real (opentelemetry) world this is what happens: 

```ts
    const span = trace
      .getTracer(spanOptions.packageInformation.name, spanOptions.packageInformation.version)
      .startSpan(name, toSpanOptions(spanOptions));
```

So this just is passed to the tracer that we get. Then in OTel it ends up as an attribute on the span:

![image](https://user-images.githubusercontent.com/753570/143148670-7f0efc4e-b666-4765-b779-c72178984bb1.png)

One idea might be to record the package inforamtion as attributes on the span and then assert that they exist or something? Not sure...

Reply:
But either way I think we can add that later - for now if we ensure that spans are closed correctly and parented correctly after the callback runs that would be 🅰️ ➕